dot/parachain/candidate-validation/worker.go

```diff
@@ -0,0 +1,93 @@
+package candidatevalidation
+
+import (
+	parachainruntime "github.com/ChainSafe/gossamer/dot/parachain/runtime"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+)
+
+// TODO(ed): figure out a better name for this that describes what it does
+type worker struct {
+	workerID    parachaintypes.ValidationCodeHash
+	instance    *parachainruntime.Instance
+	isProcessed map[parachaintypes.CandidateHash]*ValidationResult
+}
+
+type workerTask struct {
+	work             parachainruntime.ValidationParameters
+	maxPoVSize       uint32
+	candidateReceipt *parachaintypes.CandidateReceipt
+}
+
+func newWorker(validationCode parachaintypes.ValidationCode) (*worker, error) {
+	validationRuntime, err := parachainruntime.SetupVM(validationCode)
+
+	if err != nil {
+		return nil, err
+	}
+	return &worker{
+		workerID:    validationCode.Hash(),
+		instance:    validationRuntime,
+		isProcessed: make(map[parachaintypes.CandidateHash]*ValidationResult),
+	}, nil
+}
+
+func (w *worker) executeRequest(task *workerTask) (*ValidationResult, error) {
+	logger.Debugf("[EXECUTING] worker %x task %v", w.workerID, task.work)
+	candidateHash, err := parachaintypes.GetCandidateHash(task.candidateReceipt)
+	if err != nil {
+		return nil, err
+	}
+
+	// do isProcessed check here
+	if processed, ok := w.isProcessed[candidateHash]; ok {
+		logger.Debugf("candidate %x already processed", candidateHash)
```

Comment:
should this ever happen?

Reply:
It seems that this may happen, this was done to handle the requirement from host implementer's guide "Multiple requests for preparing the same artifact are coalesced, so that the work is only done once."